packages/dht/src/dht/PeerManager.ts

```diff
@@ -0,0 +1,320 @@
+            localNodeId: selfId.value,
+            numberOfNodesPerKBucket: this.config.numberOfNodesPerKBucket,
+            numberOfNodesToPing: this.config.numberOfNodesPerKBucket
+        })
+
+        this.bucket.on('ping', (oldContacts: RemoteDhtNode[], newContact: RemoteDhtNode) => this.onKBucketPing(oldContacts, newContact))
+        this.bucket.on('removed', (contact: RemoteDhtNode) => this.onKBucketRemoved(contact))
+        this.bucket.on('added', (contact: RemoteDhtNode) => this.onKBucketAdded(contact))
+        this.bucket.on('updated', (_oldContact: RemoteDhtNode, _newContact: RemoteDhtNode) => {
+            // TODO: Update contact info to the connection manager and reconnect
+        })
+        this.neighborList = new SortedContactList(selfId, this.config.maxNeighborListSize)
+        this.neighborList.on('contactRemoved', (removedContact: RemoteDhtNode, closestContacts: RemoteDhtNode[]) => {
+            if (this.stopped) {
+                return
+            }
+            this.emit('contactRemoved', removedContact.getPeerDescriptor(), closestContacts.map((peer) => peer.getPeerDescriptor()))
+            this.randomPeers!.addContact(removedContact)
+        })
+        this.neighborList.on('newContact', (newContact: RemoteDhtNode, closestContacts: RemoteDhtNode[]) =>
+            this.emit('newContact', newContact.getPeerDescriptor(), closestContacts.map((peer) => peer.getPeerDescriptor()))
+        )
+        this.openInternetPeers = new SortedContactList(selfId, this.config.maxNeighborListSize / 2)
+        this.openInternetPeers.on('contactRemoved', (removedContact: RemoteDhtNode, closestContacts: RemoteDhtNode[]) =>
+            this.emit('openInternetContactRemoved', removedContact.getPeerDescriptor(), closestContacts.map((peer) => peer.getPeerDescriptor()))
+        )
+        this.openInternetPeers.on('newContact', (newContact: RemoteDhtNode, closestContacts: RemoteDhtNode[]) =>
+            this.emit('newOpenInternetContact', newContact.getPeerDescriptor(), closestContacts.map((peer) => peer.getPeerDescriptor()))
+        )
+        
+        this.randomPeers = new RandomContactList(selfId, this.config.maxNeighborListSize)
+        this.randomPeers.on('contactRemoved', (removedContact: RemoteDhtNode, closestContacts: RemoteDhtNode[]) =>
+            this.emit('randomContactRemoved', removedContact.getPeerDescriptor(), closestContacts.map((peer) => peer.getPeerDescriptor()))
+        )
+        this.randomPeers.on('newContact', (newContact: RemoteDhtNode, closestContacts: RemoteDhtNode[]) =>
+            this.emit('newRandomContact', newContact.getPeerDescriptor(), closestContacts.map((peer) => peer.getPeerDescriptor()))
+        )
+    }
+
+    private onKBucketPing(oldContacts: RemoteDhtNode[], newContact: RemoteDhtNode): void {
+        if (this.stopped) {
+            return
+        }
+        const sortingList: SortedContactList<RemoteDhtNode> = new SortedContactList(this.config.ownPeerId!, 100)
```

Comment:
`this.config.ownPeerId` is never `undefined`

```suggestion
        const sortingList: SortedContactList<RemoteDhtNode> = new SortedContactList(this.config.ownPeerId, 100)
```

(also 4 other places in this file)